### PR TITLE
LST Oracle Adapter Contract

### PR DESCRIPTION
**Pull request by peer2f00l**
#183

### CODE REVIEW

File: common/src/oracle/price_transformer.rs [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+        match self {
+            Self::NormalizeNativeLstPrice => {
+                let scale_factor = input / Decimal::TEN_POW_24;
+
+                let price_is_negative = if price.price.0.is_negative() { -1 } else { 1 };
+                let abs_price_u128 = i128::from(price.price.0).unsigned_abs();
+                price.price.0 = price_is_negative
+                    * i64::try_from((abs_price_u128 * scale_factor).to_u128_floor()?).ok()?;
+                price.conf.0 = u64::try_from((price.conf.0 * scale_factor).to_u128_ceil()?).ok()?;
+                Some(price)
```
Comment (joshuajbouw) [inline common/src/oracle/price_transformer.rs:27]: Two issues.

* Might be a hypothetical, but `.to_128_floor` may introduce truncated precision and `to_u128_ceil` may introduce bias.
* `.ok()` could add silent failing which could result in unintended failure on `None`.

File: contract/lst-oracle/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+        };
+
+        Owner::init(&mut self_, &env::predecessor_account_id());
+
+        self_
+    }
+
+    pub fn get_oracle_id(&self) -> AccountId {
+        self.oracle_id.clone()
+    }
```
Comment (joshuajbouw) [inline contract/lst-oracle/src/lib.rs]: Rename to `oracle_id` and its best to return always as a `&AccountId`, which later someone can choose to turn into an owned value later.

File: contract/lst-oracle/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+
+    // impl Pyth:
+
+    pub fn price_feed_exists(&self, price_identifier: PriceIdentifier) -> bool {
+        self.transformers.get(&price_identifier).is_some()
+    }
+
+    pub fn list_ema_prices_no_older_than(
+        &self,
+        price_ids: Vec<PriceIdentifier>,
```
Comment (joshuajbouw) [inline contract/lst-oracle/src/lib.rs:108]: Early exit on this `Vec` would be safe, unless its always expected to have `PriceIdentifier`s.

File: contract/lst-oracle/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+                near_sdk::log!("Transforming price: {price_id}");
+                let Some(entry) = self.transformers.get(&price_id) else {
+                    env::panic_str(&format!(
+                        "No transformer associated with price ID: {price_id}",
+                    ));
+                };
+                let Some(price) = oracle_result.get(&entry.price_id) else {
+                    env::panic_str(&format!(
+                        "Mapped price ID is not in oracle result: {price_id}",
+                    ));
```
Comment (joshuajbouw) [inline contract/lst-oracle/src/lib.rs]: Just a note on the logs and panic strings, in my experience strings take up a significant amount of data in a compiled contract to the point that we had a very large reduction in contract size. When we uniformed this at Aurora and truncated to only be IDs, this greatly reduced the footprint.

File: test-utils/src/controller/lst_oracle.rs [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+        let contract = account.deploy(wasm).await.unwrap().unwrap();
+        contract
+            .call("new")
+            .args_json(json!({
+                "oracle_id": oracle_id,
+            }))
+            .transact()
+            .await
+            .unwrap()
+            .unwrap();
```
Comment (joshuajbouw) [inline test-utils/src/controller/lst_oracle.rs:43]: Transactions can fail for a number of reasons on deployment, its best to have it captured what failed.

File: test-utils/src/controller/lst_oracle.rs [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+        &self.contract
+    }
+}
+
+impl LstOracleController {
+    pub async fn deploy(account: Account, oracle_id: &AccountId) -> Self {
+        static WASM: OnceCell<Vec<u8>> = OnceCell::const_new();
+
+        let wasm = WASM
+            .get_or_init(|| get_contract("templar_lst_oracle_contract", "contract/lst-oracle"))
```
Comment (joshuajbouw) [inline test-utils/src/controller/lst_oracle.rs:31]: From what I can tell, this expects a sync closure but get_contract is likely async, which could deadlock.

**Review by joshuajbouw — CHANGES_REQUESTED** [review]
Just a couple of issues that I noted. Otherwise, looks good.

File: common/src/oracle/price_transformer.rs (hunk 1 above)
Comment (peer2f00l) [reply common/src/oracle/price_transformer.rs:27]: 1. I'm not sure we have too much of a choice here. I think the LST assets we're dealing with are _usually_ going to be used as collateral, so we should want to bias the price down (if at all), but that's not guaranteed.

    I'm thinking in the case of `price.price`, the direction of the bias is probably not as important as its consistency (i.e. I don't think it's a good idea to try to bias the transformation "cleverly," since this is just supposed to be an oracle).

    However, in the case of `price.conf`, I'm pretty sure that rounding up is the correct call, since this transformation cannot improve the resolution of the information it's transforming. (Note that `price.conf` is unsigned and represents ±error.)

    Regardless of which direction we round `price.price`, the rounding-up of `price.conf` should cover us.

1. The `.ok()` will fail only when `abs_price_u128 * scale_factor >= 2**128`. This is possible, but only in the case that `scale_factor > 2**64`, which seems exceedingly unrealistic for an LST, and I think it is OK for the calculation to fail in this case.

File: contract/lst-oracle/src/lib.rs (hunk 4 above)
Comment (peer2f00l) [reply contract/lst-oracle/src/lib.rs]: Logs are there for debugging, so I will remove them before merging, thanks for pointing this out.

Would something like #62 be what you're referring to, otherwise? We haven't had to be too worried about binary size yet, though typed errors are always nicer.

File: test-utils/src/controller/lst_oracle.rs (hunk 5 above)
Comment (peer2f00l) [reply test-utils/src/controller/lst_oracle.rs:43]: This is just in a test controller, so I'm OK with `.unwrap()`s and stuff like that. Happy to discuss if you disagree.

File: test-utils/src/controller/lst_oracle.rs (hunk 6 above)
Comment (peer2f00l) [reply test-utils/src/controller/lst_oracle.rs:31]: This is `tokio::sync::OnceCell`, not `std::sync::OnceCell`.

**Comment by peer2f00l** [comment]
Higher-level things to cover about this contract:

1. Currently it is single-owner (transferable), and only the owner can create price transformers.
2. Price transformers cannot be updated, removed, or overridden after they have been created.
3. The price IDs are arbitrarily chosen by the creator of the price transformer. (Alternatives could include a deterministic price ID function that hashes any combination of the transformer itself, the current account ID, the # of extant transformers on this contract, chain ID, etc.).

File: common/src/oracle/price_transformer.rs (hunk 1 above)
Comment (joshuajbouw) [reply common/src/oracle/price_transformer.rs:27]: 1. Had the same lines of thought running when going through it. I think it should be fine regardless, I don't necessarily see it as a red flag just a flag.
2. Sure in principle I do agree in a realistic setting its fine, but once fuzzy tests are deployed it would produce edge cases.

File: contract/lst-oracle/src/lib.rs (hunk 4 above)
Comment (joshuajbouw) [reply contract/lst-oracle/src/lib.rs]: Yeah its that. When can get to it when we get to it, thats fine. We used stuff like "ERR_<someissue>" which was fairly general. Its possible to tack in what function it happened in as well by getting the name of the method which already is written as a string in the compiled code too for more clarity.

File: test-utils/src/controller/lst_oracle.rs (hunk 5 above)
Comment (joshuajbouw) [reply test-utils/src/controller/lst_oracle.rs:43]: Oh - I missed it was a test, yeah this is fine.

**Comment by peer2f00l** [comment]
Something else to do: forward more of the Pyth API functions in case we switch which one we're using later.

File: contract/lst-oracle/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+                i += 1;
+
+                result.insert(
+                    price_id,
+                    price
+                        .clone()
+                        .and_then(|price| entry.action.apply(price, input)),
+                );
+            }
+        }
```
Comment (joshuajbouw) [inline contract/lst-oracle/src/lib.rs:185]: Changed to use built in `iter` and enumerate, `i + 1` because `i` starts at 0 with enumerate.

```suggestion
        for (i, price_id) in original_price_ids.iter().enumerate() {
            if let Some(price) = oracle_result.get(&price_id) {
                result.insert(price_id, price.clone());
            } else {
                let Some(entry) = self.transformers.get(&price_id) else {
                    env::panic_str(&format!(
                        "No transformer associated with price ID: {price_id}",
                    ));
                };
                let Some(price) = oracle_result.get(&entry.price_id) else {
                    env::panic_str(&format!(
                        "Mapped price ID is not in oracle result: {price_id}",
                    ));
                };
                let input = callback_result::<Decimal>(i + 1);

                result.insert(
                    price_id,
                    price
                        .clone()
                        .and_then(|price| entry.action.apply(price, input)),
                );
            }
        }
```

**Review by joshuajbouw — CHANGES_REQUESTED** [review]
Small change and good to go.

File: contract/lst-oracle/src/lib.rs (hunk 7 above)
Comment (peer2f00l) [reply contract/lst-oracle/src/lib.rs:185]: This is functionally different, since price IDs earlier in the list may not have instigated an additional promise, so it's not always its positional index + 1.

Comment (joshuajbouw) [reply contract/lst-oracle/src/lib.rs:185]: Ah - yeah, my mistake. Will change it now.

Comment (joshuajbouw) [reply contract/lst-oracle/src/lib.rs:185]: You are right - good point. Disregard this suggestion.

**Review by joshuajbouw — APPROVED** [review]
LGTM!
